warehouse/macaroons/caveats/__init__.py

```diff
@@ -149,6 +145,104 @@ def verify(self, request: Request, context: Any, permission: str) -> Result:
         return Success()
 
 
+class PublicPermissions(enum.IntFlag, boundary=enum.FlagBoundary.STRICT):
+    # We use 1 << N here instead of enum.auto() because the selected value is
+    # very important to us, if the value of a permission changes, then the scope
+    # of existing Macaroons will change, which would be a major security fail.
+    #
+    # Explicitly setting our values also means that we can skip values, allowing
+    # us to reserve earlier bits for permissions that are likely going to be
+    # widely used, and push less commonly used permissions to later bits.
+    #
+    # The structure of these values should basically always be 1 << N, where N
+    # is a unique integer per permission, starting with 0 and increasing from
+    # there. Effectively N is the logical ID for a given permission.
+
+    Upload = 1 << 0
+
+
+_PERMISSION_MAPPING_VERIFY = {
+    # This Maps an internal permission string to our macaroon caveat permissions
+    # when we're verifying a given set of permissions.
+    #
+    # NOTE: If an internal permission should map to multiple caveat permissions,
+    #       you can use the normal bitwise operations, the most useful of which
+    #       will likely be | to allow any of the listed permissions to match.
+    "upload": PublicPermissions.Upload,
+}
+
+
+_PERMISSION_MAPPING_EMIT = {
+    # Map a given internal permission string to a public permission.
+    #
+    # NOTE: We track this independently of the mapping for verification because
+    #       a given internal permission may be valid for multiple public
+    #       permissions, but that doesn't mean that we want to emit the full set
+    #       of public permissions.
+    #
+    #       For instance, if we have internal permissions for "create project",
+    #       "create release", and "upload a new file" as well as a public
+    #       permission for each _and_ a overall "upload" permission that
+    #       encapsulates all 3, granting someone the "upload a new file"
+    #       permission shouldn't also grant the "upload" permission but both
+    #       upload and "upload a new file" permission should work.
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** Really liking this example, helps describe the use case.
Does this mean that by playing with two lists, we could feasibly split apart `PublicPermissions.Upload` to more discrete action, or would we use a new `PublicPermissions.XXX` to denote the more fine-grained scope and emit that instead (or in addition) 🤔?